### PR TITLE
Bump shikiji, use transformers API, expose transformers API

### PR DESCRIPTION
**Pull request by blackmann**
## Changes

- Bump `shikiji` version
- Use [transformers](https://shikiji.netlify.app/guide/transformers) API. `tranforms` is deprecated.
- Expose the `transformers` API so that some [tranformer presets](https://shikiji.netlify.app/packages/transformers) or community transformers can be used.

## Testing

There's are existing tests for this and they pass.

## Docs

We'll need to update the docs to indicate that `transformers` can be supplied to `astro.config.js` > `markdown.shikiConfig.transformers`

/cc @withastro/maintainers-docs for feedback!

### CODE REVIEW

**Comment by ematipico** [comment]
@blackmann please a have at the CI, it seems TypeScript can't build the project

**Comment by blackmann** [comment]
I'm getting this error when trying to build `astro`

```
astro:build: src/core/config/schema.ts(67,14): error TS2742: The inferred type of 'AstroConfigSchema' cannot be named without a reference to '.pnpm/shikiji-core@0.9.17/node_modules/shikiji-core/dist/chunk-types.mjs'. This is likely not portable. A type annotation is necessary.
```

Any idea what it means. It's hard trying to understand what it means. 

I don't know much about how `pnpm` works; I'm unable to find the `shikiji-core` package in any of the `node_modules` though it's stated as a dependency for `shikiji`. I dunno if that's the cause.

Ideas? @ematipico

File: packages/markdown/remark/src/shiki.ts [hunk 1]
```diff
@@ -7,31 +7,29 @@ export interface ShikiHighlighter {
...
-	'#000010': 'var(--astro-code-token-string-expression)',
-	'#000011': 'var(--astro-code-token-punctuation)',
-	'#000012': 'var(--astro-code-token-link)',
+	'--astro-code-foreground': '--astro-code-color-text',
+	'--astro-code-background': '--astro-code-color-background',
 };
 const COLOR_REPLACEMENT_REGEX = new RegExp(
 	`(${Object.keys(ASTRO_COLOR_REPLACEMENTS).join('|')})`,
 	'g'
 );
```
Comment (blackmann) [inline packages/markdown/remark/src/shiki.ts:16]: We can get rid of all these if `astro-code-color-text` and `astro-code-color-background` are renamed to match.

Should that happen? Or these are deliberate renames

File: packages/markdown/remark/src/shiki.ts [hunk 2]
```diff
@@ -53,74 +51,77 @@ export async function createShikiHighlighter({
...
+							// theme.id for shiki -> shikiji compat
+							const themeName = typeof theme === 'string' ? theme : theme.name;
+							if (themeName === 'css-variables') {
+								// Replace special color tokens to CSS variables
+								visit(node as any, 'element', (child) => {
+									if (child.properties?.style) {
+										child.properties.style = replaceCssVariables(child.properties.style);
+									}
+								});
+							}
```
Comment (blackmann) [inline packages/markdown/remark/src/shiki.ts:120]: we can get rid of these lines too if we make the token names all match.

**Comment by blackmann** [comment]
```
 Tasks:    16 successful, 16 total
Cached:    9 cached, 16 total
  Time:    4m22.645s
```

All tests passing locally for me. I can't tell what the failure is with the CIs.

@ematipico please have a look¿

File: packages/markdown/remark/src/shiki.ts (hunk 1 above)
Comment (bluwy) [reply packages/markdown/remark/src/shiki.ts:16]: We should keep `astro-code-color-text` and `astro-code-color-background` for now for backwards compatibility. We could perhaps rename them in the next major

**Review by ematipico — CHANGES_REQUESTED** [review]
Blocking, as it will be part of our next minor release. We will merge it in the next merge day :)

File: packages/astro/package.json [hunk 3]
```diff
@@ -166,7 +166,7 @@
     "resolve": "^1.22.4",
     "semver": "^7.5.4",
     "server-destroy": "^1.0.1",
-    "shikiji": "^0.6.13",
+    "shikiji": "^0.9.17",
```
Comment (tieppt) [inline packages/astro/package.json:169]:
```suggestion
    "shikiji": "^0.9.18",
```
Can we upgrade the library version to the suggested version to support Angular v17? Please refer to this changelogs https://github.com/antfu/shikiji/compare/v0.9.17...v0.9.18

File: .changeset/six-scissors-worry.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+"@astrojs/markdown-remark": minor
+"astro": minor
+---
+
+Bumps shikiji and exposes a new `markdown.shikiConfig.transformers` config option. You can use this option to transform the shikiji hast (AST format of the generated HTML) to extend the features of the final HTML. See [shikiji's documentation](https://shikiji.netlify.app/guide/transformers) for more details.
```
Comment (sarah11918) [inline .changeset/six-scissors-worry.md:6]:
```suggestion
Bumps Shikiji and exposes a new `markdown.shikiConfig.transformers` config option. You can use this option to transform the Shikiji hast (AST format of the generated HTML) to extend the features of the final HTML. See [shikiji's documentation](https://shikiji.netlify.app/guide/transformers) for more details.
```
The name Shikiji is capitalized on their website.

**Review by sarah11918 — COMMENTED** [review]
If this adds a new user-configurable option, there needs to be documentation and an example showing this. Where is the config set (code snippet)? What are the values you can set, and what do they do?

I'm going from the description provided in the changeset. If this is *not* user-configurable, then "exposes a new config option" is not the right way to describe this change, because it implies that there is something you can do here!

**Review by sarah11918 — CHANGES_REQUESTED** [review]
Blocking this, because we have no documentation.

We will need:
- JSDoc content to update the [configuration reference page](https://docs.astro.build/en/reference/configuration-reference/#markdownshikiconfig) automatically
- a PR to docs to update the [Shiki configuration section](https://docs.astro.build/en/guides/markdown-content/#shiki-configuration) as appropriate.

For a minor release, not opposed to showing a code snippet of the config in the changeset if you so decide, either!

**Comment by ematipico** [comment]
@blackmann could you please help us with documentation?

**Comment by blackmann** [comment]
Yes. On it!

**Review by ematipico — APPROVED** [review]

**Review by sarah11918 — APPROVED** [review]
Final approval for docs!

**Pull request merged**